File: services/typescript/hono/src/routes/handlers/get.ts
import type { Context } from "hono";
import { type State, parseId } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function getHandler(state: State) {
  return (c: Context) => {
    const logger = getLogger();

    const id = parseId(c.req.param("id"));
    if (!id.success) {
      logger.warn({ method: c.req.method, path: c.req.path }, "Invalid path");
      return sendError(c, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "GET", path: c.req.path },
        "Task not found"
      );
      return sendError(c, AppErrors.TaskNotFound);
    }

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "GET"
      },
      "Retrieved task"
    );

    return c.json(task, 200);
  };
}

File: services/python/flask/src/routes/handlers/get.py
import structlog
from uuid import UUID
from flask import current_app, jsonify, request
from pydantic import ValidationError

from app.params import parse_id
from app.state import AppState
from app.task import Task
from routes.errors import send_error, AppErrors


def get_handler(id: str):
    logger = structlog.get_logger()

    try:
        task_id: UUID = parse_id(id)
    except ValidationError:
        logger.warn("Invalid path", method="GET", path=request.path)
        return send_error(AppErrors.InvalidPath)

    state: AppState = current_app.extensions["state"]
    task: Task | None = state.tasks.get(task_id)

    if task is None:
        logger.warn("Task not found", id=str(task_id), method="GET", path=request.path)
        return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Retrieved task",
        id=str(task_id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method="GET",
    )

    return jsonify(task.model_dump()), 200
